### PR TITLE
Started support for multiple coordinate systems on sympy.vector

### PR DESCRIPTION
**Pull request by Upabjojr**
Support for multiple coordinate systems.

`CoordSysCartesian` has been renamed `CoordSystem3D`, a new parameter has been added to its constructor, which deals with the differential distortion of non-cartesian coordinate systems.

TODO list:
- [ ] make gradient, divergence, curl independent of the coordinate system (info about coordinate system is already contained in the var argtrees).
- [x] curl
- [ ] laplacian
- [ ] write tests
- [ ] update documentation
- [ ] decide whether to use reciprocals of the currently defined _differential parameters_.
- [ ] make sure that basis is normal when generating the _differential parameters_ (i.e. metric has to be diagonal).

### CODE REVIEW

**Comment by Upabjojr** [comment]
Well, still no tests, but one may try it:

```
spherical = CoordSystem3D('spherical',
                          lambda r, theta, phi: (r * sin(theta) * cos(phi), r * sin(theta) * sin(phi), r * cos(theta)),
                          variable_names=['r', 'theta', 'phi'])

r, theta, phi = spherical.base_scalars()
dr, dtheta, dphi = spherical.base_vectors()

from sympy.vector.functions import gradient, divergence

gradient(r, spherical)
gradient(theta, spherical)
gradient(phi, spherical)
```

The generated gradient is
```
⎡    ∂_θ  │sin(θ)│⋅∂ᵩ⎤
⎢∂ᵣ, ───, ───────────⎥
⎢     r         2    ⎥
⎣          r⋅sin (θ) ⎦
```
The simplifier of SymPy is unable to further simplify the third term.

Divergence has also been implemented, but does not give the right result.

**Comment by Upabjojr** [comment]
@moorepants @srjoglekar246

f.y.i.

**Comment by Upabjojr** [comment]
The point is that *grad*, *div*, and *curl* can be derived in any coordinate system by three parameters. These three parameters can themselves be derived from the jacobian matrix of the transformation bringing the given coordinate system base scalars back to the cartesian one.

I think an additional parameter to the class *CoordSysCartesian* can account for the distortions caused by diffeomorphisms to *grad*, *div*, and *curl*. I thus renamed it *CoordSystem3D* (I put *3D* in order to avoid name-clashes with other SymPy modules).

Unfortunately, for the spherical coordinate system there's a problem in the derivation of one of these parameter, as I found no way to tell SymPy that *sin(theta)* is always positive (*theta* ranges from *0* to *pi*), subsequently, `Abs(sin(theta))` is not simplified.

**Comment by jksuom** [comment]
There are over 50000 new lines. Is there something wrong?

**Comment by Upabjojr** [comment]
Well, I did a refactor with PyCharm... I didn't realize about the new files until I committed.

**Comment by Upabjojr** [comment]
It looks like the auto-formatter reindented the whole documentation...

**Comment by Upabjojr** [comment]
fixed

File: sympy/vector/coordsysrect.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
+import types
+from sympy.simplify import simplify
 from sympy.core.basic import Basic
 from sympy.vector.scalar import BaseScalar
-from sympy import eye, trigsimp, ImmutableMatrix as Matrix, Symbol
+from sympy import eye, trigsimp, ImmutableMatrix as Matrix, Symbol, sympify, symbols, Dummy, sqrt, Derivative, S, Tuple
```
Comment (srjoglekar246) [inline sympy/vector/coordsysrect.py]: Just a small change, but maybe use two lines for this?

File: sympy/vector/tests/test_field_functions.py [hunk 2]
```diff
@@ -11,70 +11,73 @@
                                     scalar_potential_difference)
 from sympy.utilities.pytest import raises
 
-C = CoordSysCartesian('C')
-i, j, k = C.base_vectors()
-x, y, z = C.base_scalars()
-delop = C.delop
+cartesian = CoordSystem3D('C')
+spherical = CoordSystem3D('spherical', DifferentialClass.build_instance_from_lambda(
```
Comment (srjoglekar246) [inline sympy/vector/tests/test_field_functions.py]: I like what this line is going, and the way it provides flexibility. But for users who want to initialize systems quick and efficiently, maybe we should build API (atleast for commonly used systems like spherical) that can do this with less effort?

Comment (Upabjojr) [inline sympy/vector/tests/test_field_functions.py]: Yes, this is just a draft. For *cylindrical* and *spherical* coord systems we could insert the three metric parameters directly (i.e. not deriving them). For custom coordinate system, leave this derivation function to automatically determine the differential operators.

**Comment by Upabjojr** [comment]
Now the divergence works as well:

```
In [4]: r, theta, phi = symbols('r, theta, phi')

In [5]: spherical = CoordSystem3D('spherical',
   ...:                           Lambda((r, theta, phi), (1, 1/r, 1/(r*sin(theta)))),
   ...:                           variable_names=['r', 'theta', 'phi'])

In [6]: 

In [6]: r, theta, phi = spherical.base_scalars()

In [7]: dr, dtheta, dphi = spherical.base_vectors()

In [8]: divergence(r * dr, spherical)
Out[8]: 3
```

3 is indeed the correct result, as predicted by the divergence formula for spherical coordinates:
https://en.wikipedia.org/wiki/Del_in_cylindrical_and_spherical_coordinates#Del_formulae

I abolished the *DifferentialClass* class, to rather use a *Lambda* expression mapping the *base scalars* to the *base differential coefficients*.

In case someone wishes to define a new coordinate system whose *base differential coefficients* are not known, there's a derivation function:

```
In [12]: from sympy.vector.coordsysrect import build_instance_from_lambda

In [13]: build_instance_from_lambda(lambda r, theta, phi: (r * sin(theta) * cos(phi), r * sin(theta) * sin(phi), r * cos(theta)))
Out[13]: Lambda((_d1, _d2, _d3), (1, 1/_d1, Abs(sin(_d2))/(_d1*sin(_d2)**2)))
```

Next step: curl, laplacian and vector laplacian.

**Comment by Upabjojr** [comment]
Laplacian should be *div grad f*, so the algorithm is quite ready. Just have to write tests and update documentation.

**Comment by parsoyaarihant** [comment]
@Upabjojr , I am trying to implement multiple coodrinate system in my PR #11133. I am wondering why the work on this PR is not continued. Is it possible for us to work together and complete the vector module in sympy?

**Comment by Upabjojr** [comment]
> I am trying to implement multiple coodrinate system in my PR #11133. 

Thanks for your contribution.

> I am wondering why the work on this PR is not continued.

I am working on other stuff right now. This was a proposal, basically. I suggested to implement a single class for all coordinate systems, and let the coordinate systems be the instances of that class. Currently, every coordinate system has to be implemented as a new class.

> Is it possible for us to work together and complete the vector module in sympy?

Sure. My advise is to create a unique class for all coordinate systems. Have a look at _sympy.diffgeom_ to have an idea of a possible way to structure classes.

**Comment by parsoyaarihant** [comment]
@Upabjojr , since work on divergence and curl is completed I would like to add tests for whats been implemented in this PR so far. 

- Then, begin implementing laplacian for the vector module.

- If you have any tips regarding this please comment.

**Comment by Upabjojr** [comment]
> https://en.wikipedia.org/wiki/Curvilinear_coordinates#Relation_to_Lam.C3.A9_coefficients

I would rewrite this code, instead of passing as a class parameter the transformation to return the Cartesian coordinates, I would prefer to pass the Lamé coefficients to the class constructor:

https://en.wikipedia.org/wiki/Curvilinear_coordinates#Relation_to_Lam.C3.A9_coefficients

The operators are expressed as:

https://en.wikipedia.org/wiki/Curvilinear_coordinates#Differentiation

I think I use the reciprocals of these parameters in this PR.

That care not to break the old API, SymPy should not break the previous version's API.

**Comment by czgdp1807** [comment]
On current master,
```python
>>> from sympy.vector.coordsysrect import CoordSys3D
>>> spherical = CoordSys3D('spherical', lambda r, theta, phi: (r * sin(theta) * cos(phi), r * sin(theta) * sin(phi), r * cos(theta)), variable_names=['r', 'theta', 'phi'])
>>> r, theta, phi = spherical.base_scalars()
>>> dr, dtheta, dphi = spherical.base_vectors()
>>> from sympy.vector.functions import gradient, divergence
>>> gradient(r, spherical)
/home/czgdp1807ssd/sympy_project/sympy/sympy/vector/operators.py:34: SymPyDeprecationWarning: 

coord_sys parameter has been deprecated since SymPy 1.1. Use do not
use it instead. See https://github.com/sympy/sympy/issues/12884 for
more info.

  issue=12884,
⎛                                                        1                    
⎜─────────────────────────────────────────────────────────────────────────────
⎜   __________________________________________________________________________
⎜  ╱    2                   2                       2                     2   
⎝╲╱  sin (ph)⋅sin (theta_spherical) + sin (theta_spherical)⋅cos (ph) + cos (th

                                     ⎞            
─────────────────────────────────────⎟ i_spherical
_____________________________________⎟            
                  2                  ⎟            
eta_spherical) ⎠                                  
>>> gradient(theta, spherical)
⎛                                                                            1
⎜─────────────────────────────────────────────────────────────────────────────
⎜   __________________________________________________________________________
⎜  ╱            2    2                   2                               2    
⎝╲╱  r_spherical ⋅sin (phi_spherical)⋅cos (theta_spherical) + r_spherical ⋅sin

                                                                            ⎞ 
────────────────────────────────────────────────────────────────────────────⎟ 
____________________________________________________________________________⎟ 
2                               2    2                   2                  ⎟ 
 (theta_spherical) + r_spherical ⋅cos (phi_spherical)⋅cos (theta_spherical) ⎠ 

           
j_spherical
           
           
           
>>> gradient(phi, spherical)
⎛                                                         1                   
⎜─────────────────────────────────────────────────────────────────────────────
⎜   __________________________________________________________________________
⎜  ╱            2    2                   2                               2    
⎝╲╱  r_spherical ⋅sin (phi_spherical)⋅sin (theta_spherical) + r_spherical ⋅sin

                                       ⎞            
───────────────────────────────────────⎟ k_spherical
_______________________________________⎟            
2                     2                ⎟            
 (theta_spherical)⋅cos (phi_spherical) ⎠
```

**Comment by czgdp1807** [comment]
@Upabjojr Is this still needed? Are the results in above comment okay.

**Comment by Upabjojr** [comment]
Closing this issue.

**Pull request closed without merging**